bee-network/bee-autopeering/src/peer/lists.rs

```diff
@@ -0,0 +1,381 @@
+    }
+
+    pub(crate) fn set_last_verif_request_timestamp(&mut self) {
+        self.last_verif_request = time::unix_now_secs();
+    }
+
+    pub(crate) fn set_last_verif_response_timestamp(&mut self) {
+        self.last_verif_response = time::unix_now_secs();
+    }
+
+    pub(crate) fn is_verified(&self) -> bool {
+        time::since(self.last_verif_response).expect("system clock error") < VERIFICATION_EXPIRATION_SECS
+    }
+}
+
+impl fmt::Debug for PeerMetrics {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("PeerMetrics")
+            .field("verified_count", &self.verified_count)
+            .field("last_new_peers", &self.last_new_peers)
+            .field("last_verif_request", &self.last_verif_request)
+            .field("last_verif_response", &self.last_verif_response)
+            .finish()
+    }
+}
+
+/// TODO: consider using `IndexMap` for faster search.
+#[derive(Clone)]
+pub(crate) struct PeerRing<P, const N: usize>(VecDeque<P>);
+
+impl<P: AsRef<PeerId>, const N: usize> PeerRing<P, N> {
+    /// Returns `false`, if the list already contains the id, otherwise `true`.
+    ///
+    /// The newest item will be at index `0`, the oldest at index `n`.
+    pub(crate) fn insert(&mut self, item: P) -> bool {
+        if self.contains(item.as_ref()) {
+            false
+        } else {
+            if self.is_full() {
+                self.remove_oldest();
+            }
+            self.0.push_front(item);
+            true
+        }
+    }
+
+    pub(crate) fn remove_oldest(&mut self) -> Option<P> {
+        self.0.pop_back()
+    }
+
+    pub(crate) fn remove(&mut self, peer_id: &PeerId) -> Option<P> {
+        if let Some(index) = self.find_index(peer_id) {
+            self.remove_at(index)
+        } else {
+            None
+        }
+    }
+
+    pub(crate) fn remove_at(&mut self, index: usize) -> Option<P> {
+        self.0.remove(index)
+    }
+
+    pub(crate) fn contains(&self, peer_id: &PeerId) -> bool {
+        self.0.iter().any(|v| v.as_ref() == peer_id)
+    }
+
+    pub(crate) fn find_index(&self, peer_id: &PeerId) -> Option<usize> {
+        self.0.iter().position(|v| v.as_ref() == peer_id)
+    }
+
+    pub(crate) fn find(&self, peer_id: &PeerId) -> Option<&P> {
+        self.find_index(peer_id).map(|index| self.get(index)).flatten()
+    }
+
+    pub(crate) fn find_mut(&mut self, peer_id: &PeerId) -> Option<&mut P> {
+        let index = self.find_index(peer_id);
+        if let Some(index) = index {
+            self.get_mut(index)
+        } else {
+            None
+        }
+    }
+
+    pub(crate) fn get(&self, index: usize) -> Option<&P> {
+        self.0.get(index)
+    }
+
+    pub(crate) fn get_mut(&mut self, index: usize) -> Option<&mut P> {
+        self.0.get_mut(index)
+    }
+
+    pub(crate) fn get_newest_mut(&mut self) -> Option<&mut P> {
+        self.0.get_mut(0)
+    }
+
+    // NOTE: need to be atomic operations
```

Comment:
Sounds important ?

Reply:
It's not a TODO, but a note :trollface: . So, it's an explanation for the - at first glance - weird looking API `set_newest_and_get_mut`. I wanted to always remind me/us why I made it that weird, because the two operations `set_newest` and `get_mut` should be executed without any interruption possible.